File: api/models/knowledge.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"

	"github.com/lib/pq"
	"gorm.io/gorm"
)

const (
	KnowledgeInboxName       = "Inbox"
	KnowledgeStatusActive    = "active"
	KnowledgeStatusMerged    = "merged"
	KnowledgeSourceManual    = "manual"
	KnowledgeSourceAssistant = "assistant"
	KnowledgeSourceOrganizer = "organizer"
)

type KnowledgeAttributes map[string]interface{}

func (a KnowledgeAttributes) Value() (driver.Value, error) {
	if a == nil {
		return "{}", nil
	}
	value, err := json.Marshal(a)
	if err != nil {
		return nil, fmt.Errorf("marshal knowledge attributes: %w", err)
	}
	return string(value), nil
}

func (a *KnowledgeAttributes) Scan(value interface{}) error {
	if value == nil {
		*a = KnowledgeAttributes{}
		return nil
	}

	var data []byte
	switch typed := value.(type) {
	case []byte:
		data = typed
	case string:
		data = []byte(typed)
	default:
		return fmt.Errorf("scan knowledge attributes: unsupported type %T", value)
	}

	if len(data) == 0 {
		*a = KnowledgeAttributes{}
		return nil
	}
	if err := json.Unmarshal(data, a); err != nil {
		return fmt.Errorf("scan knowledge attributes: %w", err)
	}
	return nil
}

type KnowledgeTopic struct {
	ID                     string          `gorm:"primaryKey;type:uuid;default:gen_random_uuid()" json:"id"`
	UserID                 string          `gorm:"type:uuid;not null;index:idx_knowledge_topics_user_parent_name,priority:1" json:"-"`
	User                   *User           `gorm:"foreignKey:UserID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
	ParentID               *string         `gorm:"type:uuid;index:idx_knowledge_topics_user_parent_name,priority:2" json:"parentId,omitempty"`
	Parent                 *KnowledgeTopic `gorm:"foreignKey:ParentID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT" json:"-"`
	Name                   string          `gorm:"type:text;not null" json:"name"`
	NormalizedName         string          `gorm:"type:text;not null;index:idx_knowledge_topics_user_parent_name,priority:3" json:"-"`
	Description            string          `gorm:"type:text" json:"description,omitempty"`
	Depth                  int             `gorm:"not null" json:"depth"`
	PendingWriteCount      int             `gorm:"not null;default:0" json:"pendingWriteCount"`
	OrganizationDueAt      *time.Time      `gorm:"index:idx_knowledge_topics_organization,priority:1" json:"organizationDueAt,omitempty"`
	OrganizationLeaseUntil *time.Time      `gorm:"index:idx_knowledge_topics_organization,priority:2" json:"organizationLeaseUntil,omitempty"`
	LastOrganizedAt        *time.Time      `json:"lastOrganizedAt,omitempty"`
	CreatedAt              time.Time       `json:"createdAt"`
	UpdatedAt              time.Time       `json:"updatedAt"`
	DeletedAt              gorm.DeletedAt  `gorm:"index" json:"-"`
}

type KnowledgeEntry struct {
	ID         string              `gorm:"primaryKey;type:uuid;default:gen_random_uuid()" json:"id"`
	UserID     string              `gorm:"type:uuid;not null;index:idx_knowledge_entries_user_topic,priority:1;index:idx_knowledge_entries_user_kind_occurred,priority:1" json:"-"`
	User       *User               `gorm:"foreignKey:UserID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
	TopicID    string              `gorm:"type:uuid;not null;index:idx_knowledge_entries_user_topic,priority:2" json:"topicId"`
	Topic      *KnowledgeTopic     `gorm:"foreignKey:TopicID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT" json:"topic,omitempty"`
	Kind       string              `gorm:"type:text;not null;default:'note';index:idx_knowledge_entries_user_kind_occurred,priority:2" json:"kind"`
	Title      string              `gorm:"type:text;not null" json:"title"`
	Body       string              `gorm:"type:text;not null" json:"body"`
	Attributes KnowledgeAttributes `gorm:"type:jsonb;not null;default:'{}'" json:"attributes"`
	Tags       pq.StringArray      `gorm:"type:text[];not null;default:'{}'" json:"tags"`
	OccurredAt *time.Time          `gorm:"index:idx_knowledge_entries_user_kind_occurred,priority:3" json:"occurredAt,omitempty"`
	Source     string              `gorm:"type:text;not null" json:"source"`
	Status     string              `gorm:"type:text;not null;default:'active'" json:"status"`
	Version    int                 `gorm:"not null;default:1" json:"version"`
	CreatedAt  time.Time           `json:"createdAt"`
	UpdatedAt  time.Time           `json:"updatedAt"`
	DeletedAt  gorm.DeletedAt      `gorm:"index" json:"-"`
}

type KnowledgeEntryRevision struct {
	ID             string              `gorm:"primaryKey;type:uuid;default:gen_random_uuid()" json:"id"`
	EntryID        string              `gorm:"type:uuid;not null;index:idx_knowledge_revisions_entry_version,priority:1" json:"entryId"`
	UserID         string              `gorm:"type:uuid;not null;index" json:"-"`
	Version        int                 `gorm:"not null;index:idx_knowledge_revisions_entry_version,priority:2" json:"version"`
	TopicID        string              `gorm:"type:uuid;not null" json:"topicId"`
	Kind           string              `gorm:"type:text;not null;default:'note'" json:"kind"`
	Title          string              `gorm:"type:text;not null" json:"title"`
	Body           string              `gorm:"type:text;not null" json:"body"`
	Attributes     KnowledgeAttributes `gorm:"type:jsonb;not null" json:"attributes"`
	Tags           pq.StringArray      `gorm:"type:text[];not null" json:"tags"`
	OccurredAt     *time.Time          `json:"occurredAt,omitempty"`
	Status         string              `gorm:"type:text;not null" json:"status"`
	EntryDeletedAt *time.Time          `json:"-"`
	ChangedBy      string              `gorm:"type:text;not null" json:"changedBy"`
	Operation      string              `gorm:"type:text;not null" json:"operation"`
	CreatedAt      time.Time           `json:"createdAt"`
}

type KnowledgeTopicNode struct {
	KnowledgeTopic
	EntryCount int                  `json:"entryCount"`
	ChildCount int                  `json:"childCount"`
	Children   []KnowledgeTopicNode `json:"children"`
}

type KnowledgeTopicDetail struct {
	KnowledgeTopic
	EntryCount int              `json:"entryCount"`
	ChildCount int              `json:"childCount"`
	Breadcrumb []KnowledgeTopic `json:"breadcrumb"`
}

type KnowledgeSearchResult struct {
	Entry      KnowledgeEntry   `json:"entry"`
	Breadcrumb []KnowledgeTopic `json:"breadcrumb"`
}

type CreateKnowledgeTopicRequest struct {
	ParentID    *string `json:"parentId"`
	Name        string  `json:"name"`
	Description string  `json:"description"`
}

type UpdateKnowledgeTopicRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
	ParentID    *string `json:"parentId"`
	MoveToRoot  bool    `json:"moveToRoot"`
}

type CreateKnowledgeEntryRequest struct {
	TopicID    string              `json:"topicId"`
	Kind       string              `json:"kind"`
	Title      string              `json:"title"`
	Body       string              `json:"body"`
	Attributes KnowledgeAttributes `json:"attributes"`
	Tags       []string            `json:"tags"`
	OccurredAt *time.Time          `json:"occurredAt"`
}

type UpdateKnowledgeEntryRequest struct {
	ExpectedVersion   int                  `json:"expectedVersion"`
	TopicID           *string              `json:"topicId"`
	Kind              *string              `json:"kind"`
	Title             *string              `json:"title"`
	Body              *string              `json:"body"`
	Attributes        *KnowledgeAttributes `json:"attributes"`
	ReplaceAttributes bool                 `json:"replaceAttributes"`
	Tags              *[]string            `json:"tags"`
	OccurredAt        *time.Time           `json:"occurredAt"`
	ClearOccurredAt   bool                 `json:"clearOccurredAt"`
}

type UndoKnowledgeEntryRequest struct {
	ExpectedVersion int `json:"expectedVersion"`
}

type KnowledgeEntryMutation struct {
	TopicID           *string
	Kind              *string
	Title             *string
	Body              *string
	Attributes        *KnowledgeAttributes
	ReplaceAttributes bool
	Tags              *[]string
	OccurredAt        *time.Time
	ClearOccurredAt   bool
}

type KnowledgeSearchFilter struct {
	Query           string
	TopicID         string
	IncludeChildren bool
	Kind            string
	Tag             string
	OccurredFrom    *time.Time
	OccurredTo      *time.Time
	Limit           int
}

type KnowledgeAssistantTopic struct {
	ID          string `json:"id"`
	Path        string `json:"path"`
	Description string `json:"description,omitempty"`
}

type KnowledgeAssistantEntry struct {
	ID         string              `json:"id"`
	TopicID    string              `json:"topicId"`
	TopicPath  string              `json:"topicPath"`
	Kind       string              `json:"kind"`
	Title      string              `json:"title"`
	Body       string              `json:"body"`
	Attributes KnowledgeAttributes `json:"attributes"`
	Tags       []string            `json:"tags"`
	OccurredAt *time.Time          `json:"occurredAt,omitempty"`
	Version    int                 `json:"version"`
}

type KnowledgeAssistantContext struct {
	Topics  []KnowledgeAssistantTopic `json:"topics"`
	Entries []KnowledgeAssistantEntry `json:"recentEntries"`
}

type KnowledgeOrganizerTarget struct {
	ID                string `json:"id"`
	Name              string `json:"name"`
	ParentPath        string `json:"parentPath,omitempty"`
	Description       string `json:"description,omitempty"`
	DirectEntryCount  int    `json:"directEntryCount"`
	PendingWriteCount int    `json:"pendingWriteCount"`
	Depth             int    `json:"depth"`
}

type KnowledgeOrganizerChild struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	EntryCount  int    `json:"entryCount"`
	Depth       int    `json:"depth"`
}

type KnowledgeOrganizerEntry struct {
	ID         string              `json:"id"`
	TopicID    string              `json:"topicId"`
	Kind       string              `json:"kind"`
	Title      string              `json:"title"`
	Body       string              `json:"bodyExcerpt"`
	Tags       []string            `json:"tags"`
	Attributes KnowledgeAttributes `json:"attributes,omitempty"`
	OccurredAt *time.Time          `json:"occurredAt,omitempty"`
	Version    int                 `json:"version"`
	UpdatedAt  time.Time           `json:"updatedAt"`
}

type KnowledgeOrganizerRevision struct {
	EntryID   string    `json:"entryId"`
	TopicID   string    `json:"topicId"`
	Title     string    `json:"previousTitle"`
	Operation string    `json:"operation"`
	CreatedAt time.Time `json:"createdAt"`
}

type KnowledgeOrganizerContext struct {
	Target          KnowledgeOrganizerTarget     `json:"targetTopic"`
	Children        []KnowledgeOrganizerChild    `json:"childTopics"`
	RecentEntries   []KnowledgeOrganizerEntry    `json:"recentEntries"`
	Tags            []string                     `json:"existingTags"`
	SimilarEntries  []KnowledgeOrganizerEntry    `json:"similarOlderEntries"`
	RecentRevisions []KnowledgeOrganizerRevision `json:"recentOrganizerRevisions"`
}

type KnowledgeOrganizationOperation struct {
	Type                     string   `json:"type"`
	EntryID                  string   `json:"entryId,omitempty"`
	ExpectedVersion          int      `json:"expectedVersion,omitempty"`
	Tags                     []string `json:"tags,omitempty"`
	TemporaryID              string   `json:"temporaryId,omitempty"`
	Name                     string   `json:"name,omitempty"`
	Description              string   `json:"description,omitempty"`
	TargetTopicID            string   `json:"targetTopicId,omitempty"`
	Title                    string   `json:"title,omitempty"`
	CanonicalEntryID         string   `json:"canonicalEntryId,omitempty"`
	DuplicateEntryID         string   `json:"duplicateEntryId,omitempty"`
	CanonicalExpectedVersion int      `json:"canonicalExpectedVersion,omitempty"`
	DuplicateExpectedVersion int      `json:"duplicateExpectedVersion,omitempty"`
	Reason                   string   `json:"reason"`
}

type KnowledgeOrganizationPlan struct {
	Operations []KnowledgeOrganizationOperation `json:"operations"`
}

type KnowledgeOrganizationClaim struct {
	Topic                        KnowledgeTopic `json:"-"`
	LeaseUntil                   time.Time      `json:"-"`
	SyntheticManual              bool           `json:"-"`
	PendingWriteCountBeforeClaim int            `json:"-"`
	OrganizationDueAtBeforeClaim *time.Time     `json:"-"`
}

type KnowledgeOrganizationApplyResult struct {
	OperationsApplied int      `json:"operationsApplied"`
	ChangedEntryIDs   []string `json:"changedEntryIds"`
	CreatedTopicIDs   []string `json:"createdTopicIds"`
	AffectedTopicIDs  []string `json:"affectedTopicIds"`
}

type KnowledgeOrganizationResponse struct {
	Status string                           `json:"status"`
	Topic  KnowledgeTopic                   `json:"topic"`
	Result KnowledgeOrganizationApplyResult `json:"result"`
}

File: app/lib/models/knowledge.dart
const Object _knowledgeUnset = Object();

DateTime _knowledgeDate(dynamic value) {
  return DateTime.tryParse(value?.toString() ?? '') ??
      DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
}

class KnowledgeTopic {
  const KnowledgeTopic({
    required this.id,
    required this.name,
    required this.depth,
    required this.pendingWriteCount,
    required this.createdAt,
    required this.updatedAt,
    this.parentId,
    this.description = '',
    this.organizationDueAt,
    this.organizationLeaseUntil,
    this.lastOrganizedAt,
  });

  final String id;
  final String? parentId;
  final String name;
  final String description;
  final int depth;
  final int pendingWriteCount;
  final DateTime? organizationDueAt;
  final DateTime? organizationLeaseUntil;
  final DateTime? lastOrganizedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  bool get isInbox => parentId == null && name.toLowerCase() == 'inbox';

  factory KnowledgeTopic.fromJson(Map<String, dynamic> json) {
    return KnowledgeTopic(
      id: json['id']?.toString() ?? '',
      parentId: json['parentId']?.toString(),
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      depth: (json['depth'] as num?)?.toInt() ?? 0,
      pendingWriteCount: (json['pendingWriteCount'] as num?)?.toInt() ?? 0,
      organizationDueAt:
          json['organizationDueAt'] == null
              ? null
              : _knowledgeDate(json['organizationDueAt']),
      organizationLeaseUntil:
          json['organizationLeaseUntil'] == null
              ? null
              : _knowledgeDate(json['organizationLeaseUntil']),
      lastOrganizedAt:
          json['lastOrganizedAt'] == null
              ? null
              : _knowledgeDate(json['lastOrganizedAt']),
      createdAt: _knowledgeDate(json['createdAt']),
      updatedAt: _knowledgeDate(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    if (parentId != null) 'parentId': parentId,
    'name': name,
    if (description.isNotEmpty) 'description': description,
    'depth': depth,
    'pendingWriteCount': pendingWriteCount,
    if (organizationDueAt != null)
      'organizationDueAt': organizationDueAt!.toIso8601String(),
    if (organizationLeaseUntil != null)
      'organizationLeaseUntil': organizationLeaseUntil!.toIso8601String(),
    if (lastOrganizedAt != null)
      'lastOrganizedAt': lastOrganizedAt!.toIso8601String(),
    'createdAt': createdAt.toIso8601String(),
    'updatedAt': updatedAt.toIso8601String(),
  };

  KnowledgeTopic copyWith({
    String? id,
    Object? parentId = _knowledgeUnset,
    String? name,
    String? description,
    int? depth,
    int? pendingWriteCount,
    Object? organizationDueAt = _knowledgeUnset,
    Object? organizationLeaseUntil = _knowledgeUnset,
    Object? lastOrganizedAt = _knowledgeUnset,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return KnowledgeTopic(
      id: id ?? this.id,
      parentId:
          identical(parentId, _knowledgeUnset)
              ? this.parentId
              : parentId as String?,
      name: name ?? this.name,
      description: description ?? this.description,
      depth: depth ?? this.depth,
      pendingWriteCount: pendingWriteCount ?? this.pendingWriteCount,
      organizationDueAt:
          identical(organizationDueAt, _knowledgeUnset)
              ? this.organizationDueAt
              : organizationDueAt as DateTime?,
      organizationLeaseUntil:
          identical(organizationLeaseUntil, _knowledgeUnset)
              ? this.organizationLeaseUntil
              : organizationLeaseUntil as DateTime?,
      lastOrganizedAt:
          identical(lastOrganizedAt, _knowledgeUnset)
              ? this.lastOrganizedAt
              : lastOrganizedAt as DateTime?,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class KnowledgeOrganizationResult {
  const KnowledgeOrganizationResult({
    required this.operationsApplied,
    required this.changedEntryIds,
    required this.createdTopicIds,
    required this.affectedTopicIds,
  });

  final int operationsApplied;
  final List<String> changedEntryIds;
  final List<String> createdTopicIds;
  final List<String> affectedTopicIds;

  factory KnowledgeOrganizationResult.fromJson(Map<String, dynamic> json) {
    List<String> strings(String key) =>
        (json[key] as List<dynamic>? ?? const [])
            .map((value) => value.toString())
            .toList();
    return KnowledgeOrganizationResult(
      operationsApplied: (json['operationsApplied'] as num?)?.toInt() ?? 0,
      changedEntryIds: strings('changedEntryIds'),
      createdTopicIds: strings('createdTopicIds'),
      affectedTopicIds: strings('affectedTopicIds'),
    );
  }

  Map<String, dynamic> toJson() => {
    'operationsApplied': operationsApplied,
    'changedEntryIds': changedEntryIds,
    'createdTopicIds': createdTopicIds,
    'affectedTopicIds': affectedTopicIds,
  };
}

class KnowledgeOrganizationResponse {
  const KnowledgeOrganizationResponse({
    required this.status,
    required this.topic,
    required this.result,
  });

  final String status;
  final KnowledgeTopic topic;
  final KnowledgeOrganizationResult result;

  factory KnowledgeOrganizationResponse.fromJson(Map<String, dynamic> json) {
    return KnowledgeOrganizationResponse(
      status: json['status']?.toString() ?? '',
      topic: KnowledgeTopic.fromJson(
        Map<String, dynamic>.from(json['topic'] as Map),
      ),
      result: KnowledgeOrganizationResult.fromJson(
        Map<String, dynamic>.from(json['result'] as Map),
      ),
    );
  }

  Map<String, dynamic> toJson() => {
    'status': status,
    'topic': topic.toJson(),
    'result': result.toJson(),
  };
}

class KnowledgeTopicNode {
  const KnowledgeTopicNode({
    required this.topic,
    required this.entryCount,
    required this.childCount,
    this.children = const [],
  });

  final KnowledgeTopic topic;
  final int entryCount;
  final int childCount;
  final List<KnowledgeTopicNode> children;

  factory KnowledgeTopicNode.fromJson(Map<String, dynamic> json) {
    final children =
        (json['children'] as List<dynamic>? ?? const [])
            .map(
              (item) => KnowledgeTopicNode.fromJson(
                Map<String, dynamic>.from(item as Map),
              ),
            )
            .toList();
    return KnowledgeTopicNode(
      topic: KnowledgeTopic.fromJson(json),
      entryCount: (json['entryCount'] as num?)?.toInt() ?? 0,
      childCount: (json['childCount'] as num?)?.toInt() ?? children.length,
      children: children,
    );
  }

  Map<String, dynamic> toJson() => {
    ...topic.toJson(),
    'entryCount': entryCount,
    'childCount': childCount,
    'children': children.map((child) => child.toJson()).toList(),
  };

  KnowledgeTopicNode copyWith({
    KnowledgeTopic? topic,
    int? entryCount,
    int? childCount,
    List<KnowledgeTopicNode>? children,
  }) {
    return KnowledgeTopicNode(
      topic: topic ?? this.topic,
      entryCount: entryCount ?? this.entryCount,
      childCount: childCount ?? this.childCount,
      children: children ?? this.children,
    );
  }
}

class KnowledgeTopicDetail {
  const KnowledgeTopicDetail({
    required this.topic,
    required this.entryCount,
    required this.childCount,
    required this.breadcrumb,
  });

  final KnowledgeTopic topic;
  final int entryCount;
  final int childCount;
  final List<KnowledgeTopic> breadcrumb;

  factory KnowledgeTopicDetail.fromJson(Map<String, dynamic> json) {
    return KnowledgeTopicDetail(
      topic: KnowledgeTopic.fromJson(json),
      entryCount: (json['entryCount'] as num?)?.toInt() ?? 0,
      childCount: (json['childCount'] as num?)?.toInt() ?? 0,
      breadcrumb:
          (json['breadcrumb'] as List<dynamic>? ?? const [])
              .map(
                (item) => KnowledgeTopic.fromJson(
                  Map<String, dynamic>.from(item as Map),
                ),
              )
              .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
    ...topic.toJson(),
    'entryCount': entryCount,
    'childCount': childCount,
    'breadcrumb': breadcrumb.map((part) => part.toJson()).toList(),
  };
}

class KnowledgeEntry {
  const KnowledgeEntry({
    required this.id,
    required this.topicId,
    required this.kind,
    required this.title,
    required this.body,
    required this.attributes,
    required this.tags,
    required this.source,
    required this.status,
    required this.version,
    required this.createdAt,
    required this.updatedAt,
    this.occurredAt,
    this.topic,
  });

  final String id;
  final String topicId;
  final String kind;
  final String title;
  final String body;
  final Map<String, dynamic> attributes;
  final List<String> tags;
  final DateTime? occurredAt;
  final String source;
  final String status;
  final int version;
  final DateTime createdAt;
  final DateTime updatedAt;
  final KnowledgeTopic? topic;

  DateTime get displayDate => occurredAt ?? updatedAt;

  factory KnowledgeEntry.fromJson(Map<String, dynamic> json) {
    return KnowledgeEntry(
      id: json['id']?.toString() ?? '',
      topicId: json['topicId']?.toString() ?? '',
      kind: json['kind']?.toString() ?? 'note',
      title: json['title']?.toString() ?? '',
      body: json['body']?.toString() ?? '',
      attributes:
          json['attributes'] is Map
              ? Map<String, dynamic>.from(json['attributes'] as Map)
              : <String, dynamic>{},
      tags:
          (json['tags'] as List<dynamic>? ?? const [])
              .map((tag) => tag.toString())
              .toList(),
      occurredAt:
          json['occurredAt'] == null
              ? null
              : _knowledgeDate(json['occurredAt']),
      source: json['source']?.toString() ?? '',
      status: json['status']?.toString() ?? 'active',
      version: (json['version'] as num?)?.toInt() ?? 1,
      createdAt: _knowledgeDate(json['createdAt']),
      updatedAt: _knowledgeDate(json['updatedAt']),
      topic:
          json['topic'] is Map
              ? KnowledgeTopic.fromJson(
                Map<String, dynamic>.from(json['topic'] as Map),
              )
              : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'topicId': topicId,
    'kind': kind,
    'title': title,
    'body': body,
    'attributes': attributes,
    'tags': tags,
    if (occurredAt != null) 'occurredAt': occurredAt!.toIso8601String(),
    'source': source,
    'status': status,
    'version': version,
    'createdAt': createdAt.toIso8601String(),
    'updatedAt': updatedAt.toIso8601String(),
    if (topic != null) 'topic': topic!.toJson(),
  };

  KnowledgeEntry copyWith({
    String? id,
    String? topicId,
    String? kind,
    String? title,
    String? body,
    Map<String, dynamic>? attributes,
    List<String>? tags,
    Object? occurredAt = _knowledgeUnset,
    String? source,
    String? status,
    int? version,
    DateTime? createdAt,
    DateTime? updatedAt,
    Object? topic = _knowledgeUnset,
  }) {
    return KnowledgeEntry(
      id: id ?? this.id,
      topicId: topicId ?? this.topicId,
      kind: kind ?? this.kind,
      title: title ?? this.title,
      body: body ?? this.body,
      attributes: attributes ?? this.attributes,
      tags: tags ?? this.tags,
      occurredAt:
          identical(occurredAt, _knowledgeUnset)
              ? this.occurredAt
              : occurredAt as DateTime?,
      source: source ?? this.source,
      status: status ?? this.status,
      version: version ?? this.version,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      topic:
          identical(topic, _knowledgeUnset)
              ? this.topic
              : topic as KnowledgeTopic?,
    );
  }
}

class KnowledgeSearchResult {
  const KnowledgeSearchResult({required this.entry, required this.breadcrumb});

  final KnowledgeEntry entry;
  final List<KnowledgeTopic> breadcrumb;

  factory KnowledgeSearchResult.fromJson(Map<String, dynamic> json) {
    return KnowledgeSearchResult(
      entry: KnowledgeEntry.fromJson(
        Map<String, dynamic>.from(json['entry'] as Map),
      ),
      breadcrumb:
          (json['breadcrumb'] as List<dynamic>? ?? const [])
              .map(
                (item) => KnowledgeTopic.fromJson(
                  Map<String, dynamic>.from(item as Map),
                ),
              )
              .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
    'entry': entry.toJson(),
    'breadcrumb': breadcrumb.map((part) => part.toJson()).toList(),
  };
}

class KnowledgeSearchFilter {
  const KnowledgeSearchFilter({
    this.query = '',
    this.topicId,
    this.includeChildren = false,
    this.kind = '',
    this.tag = '',
    this.occurredFrom,
    this.occurredTo,
    this.limit = 30,
  });

  final String query;
  final String? topicId;
  final bool includeChildren;
  final String kind;
  final String tag;
  final DateTime? occurredFrom;
  final DateTime? occurredTo;
  final int limit;
}

class KnowledgeDirectorySelection {
  const KnowledgeDirectorySelection({
    required this.breadcrumb,
    required this.children,
    this.topic,
  });

  final KnowledgeTopicNode? topic;
  final List<KnowledgeTopic> breadcrumb;
  final List<KnowledgeTopicNode> children;
}

List<KnowledgeTopicNode> flattenKnowledgeTopics(
  List<KnowledgeTopicNode> roots,
) {
  final result = <KnowledgeTopicNode>[];
  void addNode(KnowledgeTopicNode node) {
    result.add(node);
    for (final child in node.children) {
      addNode(child);
    }
  }

  for (final root in roots) {
    addNode(root);
  }
  return result;
}

KnowledgeTopicNode? findKnowledgeTopic(
  List<KnowledgeTopicNode> roots,
  String topicId,
) {
  for (final node in flattenKnowledgeTopics(roots)) {
    if (node.topic.id == topicId) return node;
  }
  return null;
}

KnowledgeDirectorySelection selectKnowledgeDirectory(
  List<KnowledgeTopicNode> roots,
  String? topicId,
) {
  if (topicId == null) {
    return KnowledgeDirectorySelection(breadcrumb: const [], children: roots);
  }
  final node = findKnowledgeTopic(roots, topicId);
  if (node == null) {
    return const KnowledgeDirectorySelection(breadcrumb: [], children: []);
  }

  final byId = {
    for (final candidate in flattenKnowledgeTopics(roots))
      candidate.topic.id: candidate.topic,
  };
  final reversed = <KnowledgeTopic>[];
  KnowledgeTopic? current = node.topic;
  while (current != null && reversed.length < 5) {
    reversed.add(current);
    current = current.parentId == null ? null : byId[current.parentId!];
  }
  return KnowledgeDirectorySelection(
    topic: node,
    breadcrumb: reversed.reversed.toList(),
    children: node.children,
  );
}

List<KnowledgeTopicNode> rebuildKnowledgeTopicTree(
  Iterable<KnowledgeTopicNode> nodes,
) {
  final all = nodes.toList();
  final byParent = <String?, List<KnowledgeTopicNode>>{};
  for (final node in all) {
    byParent.putIfAbsent(node.topic.parentId, () => []).add(node);
  }

  int compare(KnowledgeTopicNode left, KnowledgeTopicNode right) {
    if (left.topic.isInbox != right.topic.isInbox) {
      return left.topic.isInbox ? -1 : 1;
    }
    return left.topic.name.toLowerCase().compareTo(
      right.topic.name.toLowerCase(),
    );
  }

  KnowledgeTopicNode build(KnowledgeTopicNode source, int depth) {
    final children = [...?byParent[source.topic.id]]..sort(compare);
    final builtChildren =
        children.map((child) => build(child, depth + 1)).toList();
    return source.copyWith(
      topic: source.topic.copyWith(depth: depth),
      childCount: builtChildren.length,
      children: builtChildren,
    );
  }

  final roots = [...?byParent[null]]..sort(compare);
  return roots.map((root) => build(root, 0)).toList();
}

List<KnowledgeTopicNode> replaceKnowledgeTopic(
  List<KnowledgeTopicNode> roots,
  KnowledgeTopic topic,
) {
  final flat =
      flattenKnowledgeTopics(roots)
          .map(
            (node) =>
                node.topic.id == topic.id ? node.copyWith(topic: topic) : node,
          )
          .toList();
  return rebuildKnowledgeTopicTree(flat);
}

List<KnowledgeTopicNode> removeKnowledgeTopic(
  List<KnowledgeTopicNode> roots,
  String topicId,
) {
  final flat =
      flattenKnowledgeTopics(
        roots,
      ).where((node) => node.topic.id != topicId).toList();
  return rebuildKnowledgeTopicTree(flat);
}

List<KnowledgeTopicNode> adjustKnowledgeEntryCount(
  List<KnowledgeTopicNode> roots,
  String topicId,
  int delta,
) {
  return roots
      .map(
        (node) => node.copyWith(
          entryCount:
              node.topic.id == topicId
                  ? (node.entryCount + delta).clamp(0, 1 << 31)
                  : node.entryCount,
          children: adjustKnowledgeEntryCount(node.children, topicId, delta),
        ),
      )
      .toList();
}
